posts/2021-02-11-const-generics-mvp-beta.md

```diff
@@ -0,0 +1,127 @@
+
+Even if you don't know what const generics are (in which case, read on!), you've likely been benefitting from them: const generics are already employed in the Rust compiler to improve the ergonomics of arrays and diagnostics; more on that below.
+
+With const generics hitting beta, let's take a quick look over what's actually being stabilized, what this means practically, and what's next.
+
+## What are const generics?
+
+Const generics are generic arguments that range over constant values, rather than types or lifetimes. This allows, for instance, types to be parameterized by integers. In fact, there has been one example of const generic types since early on in Rust's development: the array types `[T; N]`, for some type `T` and `N: usize`. However, there has previously been no way to abstract over arrays of an arbitrary size: if you wanted to implement a trait for arrays of any size, you would have to do so manually for each possible value. For a long time, even the standard library methods for arrays were limited to arrays of length at most 32 due to this problem. This restriction was [finally lifted in Rust 1.47](https://blog.rust-lang.org/2020/10/08/Rust-1.47.html#traits-on-larger-arrays) - a change that was made possible by const generics.
+
+Here's an example of a type and implementation making use of const generics: a type wrapping a pair of arrays of the same size.
+
+```rust
+struct ArrayPair<T, const N: usize> {
+    left: [T; N],
+    right: [T; N],
+}
+
+impl<T: Debug, const N: usize> Debug for ArrayPair<T, N> {
+    // ...
+}
+```
+
+### Current restrictions
+
+The first iteration of const generics has been deliberately constrained: in other words, this version is the MVP (minimal viable product) for const generics. This decision is motivated both by the additional complexity of general const generics (the implementation for general const generics is not yet complete, but we feel const generics in 1.51 are already very useful), as well as by the desire to introduce a large feature gradually, to gain experience with any potential shortcomings and difficulties. We intend to lift these in future versions of Rust: see [what's next](#whats-next).
+
+#### Only integral types are permitted for const generics
+
+For now, the only types that may be used as the type of a const generic argument are the types of integers (i.e. signed and unsigned integers, including `isize` and `usize`) as well as `char` and `bool`. This covers a primary use case of const, namely abstracting over arrays. In the future, this restriction will be lifted to allow more complex types, such as `&str` and user-defined types.
+
+#### No complex generic expressions in const arguments
+
+Currently, const parameters may only be instantiated by const arguments of the current forms:
```

Comment:
What does "of the current form" mean? Maybe you meant to remove that in favor of "Currently,"?

Reply:
I think they might have meant "of the following forms".

Reply:
> I think they might have meant "of the following forms".

Ah, that's much clearer. Can it be changed to that?

Reply:
```suggestion
Currently, const parameters may only be instantiated by const arguments of the following forms:
```